File: Tests/YToolsTests/BackgroundManagerPersistenceTests.swift
import Foundation
import XCTest
@testable import YTools

@MainActor
final class BackgroundManagerPersistenceTests: XCTestCase {
    func testSnippetSaveDuringLoadMergesAndFlushesLatestRevision() async throws {
        let root = try temporaryDirectory()
        defer { try? FileManager.default.removeItem(at: root) }
        let url = root.appendingPathComponent("snippets.enc")
        let key = Data(repeating: 9, count: 32)
        let store = SecureCodableStore(fileURL: url) { _ in key }
        let now = Date()
        XCTAssertTrue(store.save([
            SnippetItem(id: UUID(), title: "existing", keyword: "", content: "old", collection: "默认", createdAt: now, updatedAt: now)
        ]))

        let manager = SnippetManager { store }
        XCTAssertTrue(manager.save(text: "first", title: "first"))
        XCTAssertTrue(manager.save(text: "second", title: "second"))
        await manager.flushPendingChanges()

        guard case let .loaded(items) = store.load([SnippetItem].self) else {
            return XCTFail("Expected encrypted snippet snapshot")
        }
        XCTAssertEqual(Set(items.map(\.title)), Set(["existing", "first", "second"]))
        XCTAssertEqual(items.first?.title, "second")
    }

    func testFailedSnippetSaveKeepsNewestMemoryValue() async throws {
        let root = try temporaryDirectory()
        defer { try? FileManager.default.removeItem(at: root) }
        let url = root.appendingPathComponent("snippets.enc")
        try Data([1, 2, 3]).write(to: url)
        let store = SecureCodableStore(fileURL: url) { _ in Data(repeating: 4, count: 32) }
        let manager = SnippetManager { store }
        await manager.waitUntilLoaded()

        let saved = await manager.savePersisted(text: "newest", title: "newest")
        XCTAssertFalse(saved)
        XCTAssertTrue(manager.items.contains { $0.title == "newest" })
    }

    func testRecentDocumentRecordedDuringLoadSurvivesFlush() async throws {
        let root = try temporaryDirectory()
        defer { try? FileManager.default.removeItem(at: root) }
        let vault = root.appendingPathComponent("recent.enc")
        let old = root.appendingPathComponent("old.txt")
        let new = root.appendingPathComponent("new.txt")
        try Data("old".utf8).write(to: old)
        try Data("new".utf8).write(to: new)
        let store = SecureCodableStore(fileURL: vault) { _ in Data(repeating: 5, count: 32) }
        XCTAssertTrue(store.save([RecentDocumentItem(id: UUID(), path: old.path, lastOpenedAt: Date())]))

        let manager = RecentDocumentsManager { store }
        manager.record(new)
        await manager.flushPendingChanges()

        guard case let .loaded(items) = store.load([RecentDocumentItem].self) else {
            return XCTFail("Expected encrypted recent-document snapshot")
        }
        XCTAssertEqual(items.map(\.path), [new.path, old.path])
    }

    private func temporaryDirectory() throws -> URL {
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("ytools-manager-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: url, withIntermediateDirectories: true)
        return url
    }
}

File: Tests/YTools.Windows.Tests/BackgroundManagerPersistenceTests.cs
using System.IO;
using YTools.Models;
using YTools.Services;
using YTools.Services.Storage;

namespace YTools.Windows.Tests;

public class BackgroundManagerPersistenceTests
{
    [Fact]
    public async Task SnippetSaveDuringLoadMergesWithEncryptedSnapshotAndFlushesInOrder()
    {
        var root = CreateRoot();
        var path = Path.Combine(root, "snippets.enc");
        var key = Enumerable.Range(1, 32).Select(value => (byte)value).ToArray();
        var store = new SecureCodableStore(path, _ => key);
        var now = DateTimeOffset.UtcNow;
        Assert.True(store.Save(new List<SnippetItem>
        {
            new(Guid.NewGuid(), "existing", "", "old", "默认", now, now)
        }));

        try
        {
            var manager = new SnippetManager(() => store);
            var first = manager.SaveAsync("first", title: "first");
            var second = manager.SaveAsync("second", title: "second");

            Assert.True(await first);
            Assert.True(await second);
            await manager.FlushPendingChangesAsync();

            var loaded = store.Load<List<SnippetItem>>();
            Assert.Equal(SecureStoreLoadResultKind.Loaded, loaded.Kind);
            Assert.Equal(new[] { "second", "first", "existing" }, loaded.Value!.Select(item => item.Title));
        }
        finally
        {
            Directory.Delete(root, recursive: true);
        }
    }

    [Fact]
    public async Task FailedSnippetSaveKeepsNewestInMemoryRevision()
    {
        var root = CreateRoot();
        var path = Path.Combine(root, "snippets.enc");
        File.WriteAllBytes(path, [1, 2, 3]);
        var store = new SecureCodableStore(path, _ => new byte[32]);
        try
        {
            var manager = new SnippetManager(() => store);
            await manager.WaitUntilLoadedAsync();

            Assert.False(await manager.SaveAsync("newest", title: "newest"));
            Assert.Contains(manager.Items, item => item.Title == "newest");
        }
        finally
        {
            Directory.Delete(root, recursive: true);
        }
    }

    [Fact]
    public async Task RecentDocumentRecordedDuringLoadIsIncludedAtFlush()
    {
        var root = CreateRoot();
        var vault = Path.Combine(root, "recent.enc");
        var oldFile = Path.Combine(root, "old.txt");
        var newFile = Path.Combine(root, "new.txt");
        File.WriteAllText(oldFile, "old");
        File.WriteAllText(newFile, "new");
        var key = Enumerable.Repeat((byte)7, 32).ToArray();
        var store = new SecureCodableStore(vault, _ => key);
        Assert.True(store.Save(new List<RecentDocumentItem>
        {
            new(Guid.NewGuid(), oldFile, DateTimeOffset.UtcNow)
        }));

        try
        {
            var manager = new RecentDocumentsManager(() => store);
            manager.Record(newFile);
            await manager.FlushPendingChangesAsync();

            var loaded = store.Load<List<RecentDocumentItem>>();
            Assert.Equal(SecureStoreLoadResultKind.Loaded, loaded.Kind);
            Assert.Equal(new[] { newFile, oldFile }, loaded.Value!.Select(item => item.Path));
        }
        finally
        {
            Directory.Delete(root, recursive: true);
        }
    }

    private static string CreateRoot()
    {
        var root = Path.Combine(Path.GetTempPath(), $"ytools-manager-{Guid.NewGuid():N}");
        Directory.CreateDirectory(root);
        return root;
    }
}
